### PR TITLE
Using WireCell in standard fcl workflow by default

### PR DESCRIPTION
**Pull request by lynnt20**
This PR will fully integrate the WireCell 2D TPC Simulation/Signal Processing chain into SBND's "standard" fcl workflow. With this updated workflow, there is no longer a separate WireCell fcl/stage; the WireCell stage is merged into the standard detsim stage. This PR follows PR #409 , PR #408, and PR #414. 

Note that the drift simulation, TPC electronics simulation, and deconvolution all happen within the _detsim_ stage now. Previously the 1d drift sim (simdrift) was run in g4, the electronics sim (daq) was run in detsim, and 1d deconvolution (caldata) was run in reco1. 

A fix for space charge effects with 2D sim is now implemented by @jzennamo, where the last commit from PR #411 is cherry-picked into this PR. The 2D "fix" space charge service should be called on all modules upstream of/including detsim that use sce. 

`sbndcode/JobConfigurations/README.md` has been updated with the most recent "core" fcl workflow.

# Table of contents
1. [Standard/common fcls](#standard)
2. [2D SCE Fix/fcls](#sce)
3. [Detvar fcls](#detvar)
4. [WireCell fcls](#wirecell)
5. [Commissioning fcls](#commissioning)

## Changes to standard/common fcls <a name="standard"></a>
- `standard_g4_sbnd.fcl`: simdrift no longer run! mcreco moved to end of detsim
  - WireCell requires that ionandscint products are *not* dropped in the g4 stage. 
  - `g4_dirt_filter.fcl` (the standard g4 fcl for rockbox) no longer runs simdrift
  - `g4_simphotontime_filter.fcl` (the standard g4 fcl for in-time cosmic) has all references to simdrift/mcreco removed

- `standard_detsim_sbnd.fcl`: runs WireCell 2D Sim/SigProc
  - will output `sim::SimChannel` and `recob::Wire` by default. Configuration allowing`raw::RawDigit`s to be saved available in the commissioning fcl folder (see below).
  -  note: the electron drift _simulation_ is in the _detsim_ stage, instead of g4. 
  - note: mcreco is run at the end of detsim, since it requires `SimChannels`
 
- `reco_sbnd.fcl`: caldata no longer run! (1D deconvolution now obsolete)
  - gaushit and caloskim (calibration ntuple module) have labels updated to WireCell input
  - fasthit is now disabled in reco1, this is because we are not saving `raw::RawDigit`s by default. Additionally, no other downstream modules were using fasthit output. 

- `cafmakerjob_sbnd.fcl`: SBN-level tools CNNID and PID BDTs (Razzle/Dazzle) use WireCell inputs for WireLabel/SimChannelLabel.

PR #409 takes care of BackTracker services for truth matching. 

Fcls created to handle the alternate WireCell workflow for the 2023B production are removed in this PR, such as `g4_sce_dirt_filter_lite_wc.fcl`, `detsim_sce_lite_wc.fcl`, `reco1_sce_lite_wc2d.fcl`, and `reco1_sce_lite_wc1d.fcl.`

## Fcls Updated to 2D SCE Map <a name="sce"></a>
All the sce fcls changed were cross-referenced with the SBN Production Requests database; in other words, fcls not commonly used/not used in production were ignored.
### gen stage
- `prodcorsika_proton_intime_filter_sce.fcl` (in-time cosmic gen)
- `prodoverlay_corsika_cosmics_proton_genie_rockbox_sce.fcl` (rockbox gen)
### g4 stage 
- `g4_sce.fcl`
- `g4_sce_simphotontime_filter.fcl` (in-time cosmic g4)
  - in addition to the sce map update, a line has been added to merge the ionandscint intime and outtime products _after_ sce has been applied. This is necessary for wirecell to access the ionandscint products with sce effects applied. 
- `g4_sce_dirt_filter.fcl` (rockbox g4)
- `g4_mu_frontbackcrt_filter_sce.fcl` and `g4_mu_eastwestcrt_filter_sce.fcl` (commissioning cosmics g4 filters).
- `g4_sce_SaveCosmicMCReco.fcl` 
### detsim stage 
- `detsim_sce.fcl`
- `detsim_comm_sce_sbnd.fcl` (commissioning detsim) 

## Changes to WireCell fcls  <a name="wirecell"></a>
- `sbndcode/WireCell/wcsimsp_sbnd.fcl`: new configuration fcl to streamline referencing the WireCell sim/sp module. 
- `sbndcode/WireCell/wirecell_sim_rawdigits_sp.fcl` and `sbndcode/WireCell/wirecell_sim_sp_sbnd.fcl` now include `wcsimsp_sbnd.fcl` by default. These fcls should no longer be used as part of the "standard" workflow. They can be used if one wants to run 2D TPC Sim/SigProc only. 
### For WireCell experts:
- change the default larwirecell `SimChannel` sink module to DepoFluxWriter from the current (deprecated) *SimChannelSink modules.  A new jsonnet has been added that uses DepoFluxWriter. Default is to save the track_id from the original SimEnergyDeposits for the ML workflow.
  - Original PR about DepoFluxWriter:  LArSoft/larwirecell#35
  - SBND Validation on DepoFluxWriter: LArSoft/larwirecell#41

## Changes to DetVar fcls <a name="detvar"></a>
- All recombination fcls in recomb_variations folder are modified to `#include "g4_sce_dirt_filter_lite.fcl"`, and renamed to remove `wc` from the fcl file name. 
- All fcls in detvar_downstream folder are modified to include the standard fcls, and renamed to remove `wc` from the fcl file name.
- The diffusion fcls do not require modification since they `#include wirecell_sim_sp_sbnd.fcl`, but should be updated to include the standard detsim fcl depending on the detvar workflow. @ibsafa, happy to have feedback about this! 

## Changes to Commissioning fcls <a name="commissioning"></a>
- `detsim_comm_sbnd.fcl` includes lines to allow `raw::RawDigit`s in artroot output. 
- fasthit in hitdumper/muon track fcls now specify WireCell products as input.

### CODE REVIEW

File: sbndcode/JobConfigurations/base/reco_sbnd.fcl [hunk 1]
```diff
@@ -202,10 +202,10 @@ physics:
          , opflashtpc1
          , opflashtpc0xarapuca
          , opflashtpc1xarapuca
-         , caldata
+        #  , caldata
          , gaushit
-         , fasthit
-   #         , linecluster
+        # , fasthit
+        # , linecluster
```
Comment (henrylay97) [inline sbndcode/JobConfigurations/base/reco_sbnd.fcl:208]: Think these, and all other equivalents should be removed but that's one for @absolution1's PR.

File: sbndcode/WireCell/wcsimsp_sbnd.fcl
Comment (henrylay97) [inline sbndcode/WireCell/wcsimsp_sbnd.fcl:1]: Thanks so much for making this a PROLOG and then inheriting it :heart_eyes:

**Review by henrylay97 — APPROVED** [review]
@lynnt20 as you know I approve down to the ground of putting this PR in. You've done it in a very clean and neat way too, making all our lives easier going forward, thank you! I essentially have no comments, so unless any other wirecell people have comments on the exact configuring of the standalone wirecell fcls then it's good to go from my point of view.

File: sbndcode/JobConfigurations/base/reco_sbnd.fcl (hunk 1 above)
Comment (lynnt20) [reply sbndcode/JobConfigurations/base/reco_sbnd.fcl:208]: agreed!

File: sbndcode/WireCell/wcsimsp_sbnd.fcl
Comment (ebelchio12) [inline sbndcode/WireCell/wcsimsp_sbnd.fcl:1]: Nice to keep this single fcl

**Review by ebelchio12 — APPROVED** [review]

**Comment by bear-is-asleep** [comment]
I feel a bit uneasy about moving `mcreco` from g4 to the detsim stage, since it uses only truth information from geant4, it's natural location would be this `g4` stage. https://github.com/SBNSoftware/sbndcode/pull/409#discussion_r1491743704

**Comment by bear-is-asleep** [comment]
> I feel a bit uneasy about moving `mcreco` from g4 to the detsim stage, since it uses only truth information from geant4, it's natural location would be this `g4` stage. [#409 (comment)](https://github.com/SBNSoftware/sbndcode/pull/409#discussion_r1491743704)

You answered my question [here](https://github.com/SBNSoftware/sbndcode/pull/399#issuecomment-1961585854) thanks!

**Comment by fjnicolas** [comment]
@lynnt20 can you update these two as well?

- [fast_simulation_reconstruction_test_electrons.sh](https://github.com/SBNSoftware/sbndcode/tree/develop/test/JobConfigurations/fast_simulation_reconstruction_test_electrons.sh#L11)
- [fast_simulation_reconstruction_test_muons.sh](https://github.com/SBNSoftware/sbndcode/tree/develop/test/JobConfigurations/fast_simulation_reconstruction_test_muons.sh#L11)

File: sbndcode/JobConfigurations/standard/g4/g4_simphotontime_filter.fcl [hunk 2]
```diff
@@ -3,30 +3,31 @@
...
 # This runs the new, refactored, LArG4 simulation.
+# Note: All references to simdrift have been removed in the switch to the 2D WireCell Drift Simulation.
 
 #include "mergesimsources_sbnd.fcl"
 
 #include "standard_g4_sbnd.fcl"
 
 # Add process to finish the electron drift simulation for the intimes
-physics.producers.simdriftintime: @local::sbnd_simdrift
+# physics.producers.simdriftintime: @local::sbnd_simdrift
```
Comment (henrylay97) [inline sbndcode/JobConfigurations/standard/g4/g4_simphotontime_filter.fcl:13]: Is there much point keeping these commented out?

File: sbndcode/SpaceCharge/SpaceChargeSBND.cxx [hunk 3]
```diff
@@ -213,7 +214,15 @@ geo::Vector_t spacecharge::SpaceChargeSBND::GetPosOffsets(geo::Point_t const& po
...
       int corr = 1;
-      if (xx < 0) { corr = -1; }
+
+      if(f_2D_drift_sim_hack == true)
+	corr = -1; 
+
+      if (xx < 0) { 
+	corr = -1; 
+      }
+
```
Comment (henrylay97) [inline sbndcode/SpaceCharge/SpaceChargeSBND.cxx:224]: Is it worth adding a nice obvious comment here to try and make sure we remove it properly in future? The more places we highlight, the more likely it's properly caught.

```suggestion

      // ========================================================
      // This hack is to account for a known issue with the 
      // space charge implementation for the 2D simulation.
      // See https://cdcvs.fnal.gov/redmine/issues/28099
      // This should be removed once the appropriate upgrades
      // have been implemented.
      // ========================================================
      if(f_2D_drift_sim_hack == true)
	corr = -1; 

      if (xx < 0) {
	corr = -1; 
      }
```

**Review by henrylay97 — APPROVED** [review]
Yep, still happy with this. A couple of aesthetic comments that can be ignored. And I can't really comment on the new WireCell bit... new jsonnet etc

File: sbndcode/Commissioning/fcls/detsim_comm_sce_sbnd.fcl [hunk 4]
```diff
@@ -1,3 +1,3 @@
 #include "detsim_comm_sbnd.fcl"
 
-#include "enable_spacecharge_services_sbnd.fcl"
+#include "enable_spacecharge_services_sbnd_2D_DriftSimOnly.fcl"
```
[the remark this answers is not in the document]
Comment (lynnt20) [reply sbndcode/Commissioning/fcls/detsim_comm_sce_sbnd.fcl:3]: I had the same thought originally.. but my understanding is that we need the `2D_DriftSimOnly` only for gen, g4, and detsim stages in the 2d workflow, and the reco/caf stages still call the original sce fcl even if they are using the 2d drift sim output. So you still need two separate SCE fcls for the single 2D workflow

File: sbndcode/JobConfigurations/standard/g4/g4_simphotontime_filter.fcl (hunk 2 above)
Comment (lynnt20) [reply sbndcode/JobConfigurations/standard/g4/g4_simphotontime_filter.fcl:13]: I could clean it up and remove it entirely; I'm not that familiar with this fcl so I erred on the side of caution. Happy to just remove the lines if that makes more sense!!

File: sbndcode/SpaceCharge/SpaceChargeSBND.cxx (hunk 3 above)
Comment (lynnt20) [reply sbndcode/SpaceCharge/SpaceChargeSBND.cxx:224]: I like this! I'll commit it.

File: sbndcode/Commissioning/fcls/detsim_comm_sce_sbnd.fcl (hunk 4 above)
[the remark this answers is not in the document]
Comment (henrylay97) [reply sbndcode/Commissioning/fcls/detsim_comm_sce_sbnd.fcl:3]: Oh of course, makes perfect sense. Definitely a leave as is then :D

**Comment by lynnt20** [comment]
Some minor things to wait on before we push to develop (will update this comment as things go through)

- @henrylay97 is running some sanity check tests on the calibration ntuples to see if `pandoraSCE` successfully applies the SCE corrections as expected after the 2D SCE fix has been implemented. 
- confirmation from @bear-is-asleep that the updates to WireCell configuration play nice with the ML workflow

In the meantime I will trigger the build tests to see if there's anything obvious that still needs to be changed.

**Comment by lynnt20** [comment]
trigger build

**Comment by lynnt20** [comment]
@miquelnebot looks like all the build fails for this PR is due to a build error in sbncode associated with Cluster3D (ML workflow, also @bear-is-asleep). I traced it back to this PR: SBNSoftware/sbncode#388, where it also failed the build CI tests.

**Comment by bear-is-asleep** [comment]
> @miquelnebot looks like all the build fails for this PR is due to a build error in sbncode associated with Cluster3D (ML workflow, also @bear-is-asleep). I traced it back to this PR: [SBNSoftware/sbncode#388](https://github.com/SBNSoftware/sbncode/pull/388), where it also failed the build CI tests.

So strange because the variable it's complaining about is clearly being used [here](https://github.com/SBNSoftware/sbncode/blob/1bd88bce24935392acc107e9f4f2d69b4a8dc0b4/sbncode/Cluster3D/Cluster3DSBN_module.cc#L946-L994)

**Comment by fjnicolas** [comment]
> Looks ready to go from my point of view, other than CI approval. The clang build error is bizarre, as @bear-is-asleep points out, the variable _is_ used....
> 
> Have we run the sequential tests on this PR? I think given how careful Lynn has been with not making new fcls they should run okay, just with mega product differences?

`c14` can be really picky... the variable is indeed not used (it just assigns a value to it, but the modules doesn't further use the variable for anything, e.g. printout, save to TTree, as input for other calculations...). A patch PR will be needed in `sbncode`.

I'll trigger the sequencial tests mainly to check it runs smoothly over the different stages, though I don't expect the comparison with the old refs very meaningful as you said...

**Comment by henrylay97** [comment]
> c14 can be really picky... the variable is indeed not used (it just assigns a value to it, but the modules doesn't further use the variable for anything, e.g. printout, save to TTree, as input for other calculations...). A patch PR will be needed in sbncode.

...and this is why you're release manager mate... :eagle: :eye: 

> I'll trigger the sequencial tests mainly to check it runs smoothly over the different stages, though I don't expect the comparison with the old refs very meaningful as you said...

Completely agreed, seeing it runs through is worth it I think, on something this big & core.

**Comment by henrylay97** [comment]
As discussed on slack: this is good to go :D

**Pull request merged**